### PR TITLE
Add compound prediction mode variants for drl=2 and drl=3.

### PR DESCRIPTION
**Pull request by tdaede**

### CODE REVIEW

File: src/context.rs [hunk 1]
```diff
@@ -3529,7 +3518,23 @@ impl<'a> ContextWriter<'a> {
     };
 
     assert!(mode >= PredictionMode::NEAREST_NEARESTMV);
-    let val = mode as u32 - PredictionMode::NEAREST_NEARESTMV as u32;
+    let val = match mode {
```
Comment (lu-zero) [inline src/context.rs:3521]: what about making a small method `to_u8` or such?

File: src/stats.rs [hunk 2]
```diff
@@ -21,9 +21,9 @@ pub struct EncoderStats {
   /// Stores count of pixels belonging to each transform type in this frame
   pub tx_type_counts: [usize; TX_TYPES],
   /// Stores count of pixels belonging to each luma prediction mode in this frame
-  pub luma_pred_mode_counts: [usize; PREDICTION_MODES],
+  pub luma_pred_mode_counts: Vec<usize>,
   /// Stores count of pixels belonging to each chroma prediction mode in this frame
-  pub chroma_pred_mode_counts: [usize; PREDICTION_MODES],
+  pub chroma_pred_mode_counts: Vec<usize>,
```
Comment (lu-zero) [inline src/stats.rs:26]: why this?

File: src/context.rs (hunk 1 above)
Comment (tdaede) [reply src/context.rs:3521]: can't derive #[Debug] for arrays larger than 32

File: src/stats.rs (hunk 2 above)
Comment (tdaede) [reply src/stats.rs:26]: can't derive #[Debug] for arrays longer than 32

File: src/context.rs (hunk 1 above)
Comment (tdaede) [reply src/context.rs:3521]: could but this is the only case where this numeric conversion is used

File: src/stats.rs (hunk 2 above)
Comment (tdaede) [reply src/stats.rs:26]: I don't know why this derives debug anyway, maybe it shouldn't

Comment (lu-zero) [reply src/stats.rs:26]: Probably Debug can be dropped.

File: src/context.rs (hunk 1 above)
Comment (lu-zero) [reply src/context.rs:3521]: Ok.

**Comment by shssoichiro** [comment]
Does this impact metrics? If so, AWCY incoming?

**Comment by shssoichiro** [comment]
Also, this seems to depend on #2237? There's a lot of overlap, at least. Is that correct?

**Comment by tdaede** [comment]
No, #2237 is the one that impacts metrics.

File: src/stats.rs (hunk 2 above)
Comment (tdaede) [reply src/stats.rs:26]: Unfortunately can't do that as Packet also derives Debug.

File: src/predict.rs [hunk 3]
```diff
@@ -61,9 +61,11 @@ pub static RAV1E_INTER_COMPOUND_MODES: &[PredictionMode] = &[
   PredictionMode::NEW_NEWMV,
   PredictionMode::NEAREST_NEWMV,
   PredictionMode::NEW_NEARESTMV,
-  PredictionMode::NEAR_NEARMV,
+  PredictionMode::NEAR_NEAR0MV,
```
Comment (shssoichiro) [inline src/predict.rs:64]: Should this be updated to add the new modes?

File: src/rdo.rs [hunk 4]
```diff
@@ -1114,7 +1114,7 @@ fn inter_frame_rdo_mode_decision<T: Pixel>(
           [MotionVector::default(); 2]
         }
       }
-      PredictionMode::NEAR0MV | PredictionMode::NEAR_NEARMV => {
+      PredictionMode::NEAR0MV | PredictionMode::NEAR_NEAR0MV => {
```
Comment (shssoichiro) [inline src/rdo.rs:1117]: Should this be updated to add the new modes?

Comment (barrbrain) [reply src/rdo.rs:1117]: When the new modes are added to RAV1E_INTER_COMPOUND_MODES in #2237, the logic for those modes will be added here.

File: src/predict.rs (hunk 3 above)
Comment (barrbrain) [reply src/predict.rs:64]: The new modes will be added in #2237, which extends the search. This PR is for preparing the accounting.

File: Cargo.toml [hunk 5]
```diff
@@ -73,7 +73,7 @@ v_frame = { version = "0.1", path = "v_frame/" }
 av-metrics = { version = "0.4", optional = true }
 rayon = "1.0"
 toml = { version = "0.5", optional = true }
-arrayvec = "0.5"
+arrayvec = { version = "0.5", features = ["serde", "array-sizes-33-128"] }
```
Comment (barrbrain) [inline Cargo.toml:76]: Nice, this resolves my concern about introducing new allocations.

**Pull request merged**